analyzers/src/SonarAnalyzer.CSharp/LiveVariableAnalysis/CSharp/CSharpLiveVariableAnalysis.cs

```diff
@@ -35,61 +34,62 @@ public sealed class CSharpLiveVariableAnalysis : AbstractLiveVariableAnalysis
         private readonly ISymbol declaration;
         private readonly SemanticModel semanticModel;
 
-        private CSharpLiveVariableAnalysis(IControlFlowGraph controlFlowGraph, ISymbol declaration,
-            SemanticModel semanticModel)
-            : base(controlFlowGraph)
+        private CSharpLiveVariableAnalysis(IControlFlowGraph controlFlowGraph, ISymbol declaration, SemanticModel semanticModel) : base(controlFlowGraph)
         {
             this.declaration = declaration;
             this.semanticModel = semanticModel;
         }
 
-        public static AbstractLiveVariableAnalysis Analyze(IControlFlowGraph controlFlowGraph, ISymbol declaration,
-            SemanticModel semanticModel)
+        public static AbstractLiveVariableAnalysis Analyze(IControlFlowGraph controlFlowGraph, ISymbol declaration, SemanticModel semanticModel)
         {
             var lva = new CSharpLiveVariableAnalysis(controlFlowGraph, declaration, semanticModel);
             lva.PerformAnalysis();
             return lva;
         }
 
-        protected override void ProcessBlock(Block block, out HashSet<ISymbol> assignedInBlock,
-            out HashSet<ISymbol> usedInBlock)
+        internal static bool IsOutArgument(IdentifierNameSyntax identifier) =>
+            identifier.GetFirstNonParenthesizedParent() is ArgumentSyntax argument && argument.RefOrOutKeyword.IsKind(SyntaxKind.OutKeyword);
+
+        internal static bool IsLocalScoped(ISymbol symbol, ISymbol declaration)
         {
-            ProcessBlockInternal(block, null, out assignedInBlock, out usedInBlock);
+            return IsLocalOrParameterSymbol()
+                && symbol.ContainingSymbol != null
+                && symbol.ContainingSymbol.Equals(declaration);
+
+            bool IsLocalOrParameterSymbol() =>
+                symbol is ILocalSymbol || (symbol is IParameterSymbol parameter && parameter.RefKind == RefKind.None);
         }
 
-        private void ProcessBlockInternal(Block block, HashSet<ISymbol> processedLocalFunctions
-            , out HashSet<ISymbol> assignedInBlock, out HashSet<ISymbol> usedInBlock)
+        protected override State ProcessBlock(Block block)
         {
-            assignedInBlock = new HashSet<ISymbol>(); // Kill (The set of variables that are assigned a value.)
-            usedInBlock = new HashSet<ISymbol>(); // Gen (The set of variables that are used before any assignment.)
-
-            var assignmentLhs = new HashSet<SyntaxNode>();
+            var ret = new State();
+            ProcessBlockInternal(block, ret);
+            return ret;
+        }
 
+        private void ProcessBlockInternal(Block block, State state)
```

Comment:
wouldn't it be better for this to return `State` as well (to avoid modifying the parameter)?

You could still leave the other methods as they are, as an initial baby step

Reply:
It's part of recursive loop filling the values, so it's not easy to do just return without some heavy coping of values.